### PR TITLE
scripts: don't refresh the pacman repository on Arch

### LINKED ISSUE #12186: Installer script automatically does a partial upgrade on Arch Linux.
### What is the issue?

Arch Linux _heavily_ discourages the use of `pacman -Sy` due to the fact that this refreshes the package repositories and could cause newer libraries to be installed. This is a huge issue due to the fact that tailscale may update dependencies and break other applications depending on this automatically without any user intervention and could be huge source of headache for users of `pacman`-based distros.

Tailscale's installer should only install the package and should not be updating the repository information on the system nor upgrading any other packages.

At present, [the Tailscale automatic installer does a partial upgrade and this should not happen ever](https://github.com/tailscale/tailscale/blob/13e13555461a9b5490e97af875751ad14fb209a7/scripts/installer.sh#L516).

### Steps to reproduce

1. Install Tailscale using `install.sh`

### Are there any recent changes that introduced the issue?

_No response_

### OS

Linux

### OS version

Arch Linux

### Tailscale version

latest

### Other software

_No response_

### Bug report

BUG-594969113a09ccd7f3c3109e4e29b3fefd6d4177e423a3513a8c02f6d6ade868-20240519004900Z-bd878ed37aceaa17

### PR DESCRIPTION
**Pull request by noncombatant**
Fixes #12186

### CODE REVIEW

File: scripts/installer.sh [hunk 1]
```diff
@@ -513,7 +513,10 @@ main() {
 			;;
 		pacman)
 			set -x
-			$SUDO pacman -Sy
+			# https://github.com/tailscale/tailscale/issues/12186
+			if [ "$OS" != "arch" ]; then
+				$SUDO pacman -Sy
+			fi
```
Comment (awly) [inline scripts/installer.sh:519]: Sounds like the issue is not specific to Arch, but any distro using pacman.
We should probably never run `pacman -Sy`.

Comment (SylvKT) [reply scripts/installer.sh:519]: Why does it need to use `pacman -Sy` anyway?

**Comment by SylvKT** [comment]
The Arch CI is broken because it does not already have the package database downloaded like a regular system would. `pacman -S` and `pacman -Syu` are safe, but `pacman -Sy` is not.

File: scripts/installer.sh (hunk 1 above)
Comment (awly) [reply scripts/installer.sh:519]: It's similar to all other repos - refresh the index to fetch the latest version. Arch is the only one that actually discourages this practice.

It matters more on rpm and deb-based distros, where we own the repo and add it in this same installer script. Without a refresh, apt/dnf/yum wouldn't know about the tailscale package.

I agree that it is probably not necessary on pacman-based distros, unless you have a fresh install where `pacman -Sy` has never ran before (e.g. when you run a fresh `archlinux` docker container`).

**Comment by awly** [comment]
@noncombatant we need to update the CI config to special-case `archlinux` like we do here with others https://github.com/tailscale/tailscale/blob/061498860b10d7ffa9ad7d1a19aa554f28c15bd4/.github/workflows/installer.yml#L70-L77

**Comment by knyar** [comment]
I've just checked, and the installer script will fail on a fresh `archlinux:latest` image, with the error message advising to run `pacman -Sy`:

```
# bash install.sh
Installing Tailscale for arch, using method pacman
+ pacman -S tailscale --noconfirm
warning: database file for 'core' does not exist (use '-Sy' to download)
warning: database file for 'extra' does not exist (use '-Sy' to download)
error: target not found: tailscale
```

If it's not always safe to run that command, I think we have two options:
1. Only run the command when we need to (e.g. when database files are missing), assuming there is a stable way to determine that.
2. Completely remove the command from the installer script, and run it on CI before the script. This would not be my first choice, since ideally we'd want the script to do whatever is necessary (but not more) to install Tailscale.

**Comment by noncombatant** [comment]
We could do `if !pacman -S tailscale ...; pacman -Sy ...` (it does return with status 1 when it fails — for any reason, though, not just unknown packages), but I think that still might surprise people who don't want `-Sy` to happen without their intent. So @knyar 's option 2 might make CI most like what Arch users (and other `pacman`-users?) expect, which seems better.

**Review by knyar — APPROVED** [review]

File: .github/workflows/installer.yml [hunk 2]
```diff
@@ -67,6 +67,13 @@ jobs:
       image: ${{ matrix.image }}
       options: --user root
     steps:
+    - name: install dependencies (pacman)
+      # Refresh the package databases to ensure that the tailscale package is
+      # defined. Also, tar and gzip are needed by the actions/checkout below.
+      run: |
+        pacman -Sy
+        pacman -S tar gzip --noconfirm
```
Comment (awly) [inline .github/workflows/installer.yml:75]:
```suggestion
      # defined.
      run: pacman -Sy
```

**Review by awly — APPROVED** [review]

**Pull request merged**